### PR TITLE
Certik SwapRouter Audit Fixes 

### PR DESCRIPTION
**Pull request by defcon022**
## Description

This PR contains the fixes for SwapRouter:
- PLS-01: INCORRECT FEE AMOUNT WILL CAUSE ALL SWAPS TO FAIL
- PLS-02: INCOMPLETE CHECK
- SRS-02: MISSING CHECKS
- SRS-03: MISSING ZERO ADDRESS VALIDATION
- SWA-01: POTENTIAL REENTRANCY (OUT-OF-ORDER EVENTS)
- RHS-01: CAN USE safeTransfer()
- SRS-04: COMMENTS FOR FUNCTIONS SUPPORTING FEE
- SRS-05: COMMENTS MISSING PARAMETERS
- SRS-06: INCORRECT COMMENT
- PLS-03: INEFFICIENT CHECKS
- SRS-07: CAN USE SINGLE ADDRESS INPUT
- SWA-02: UNCHECKED BLOCKS CAN OPTIMIZE CONTRACT
- SWA-03: CUSTOM ERRORS CAN BE USED

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/Swap/SwapRouter.sol [hunk 1]
```diff
@@ -648,19 +736,22 @@ contract SwapRouter is Ownable2Step, RouterHelper, IPancakeSwapV2Router {
...
         address to,
         uint256 deadline
-    ) external virtual override ensure(deadline) ensurePath(path) returns (uint256[] memory amounts) {
+    ) external virtual override nonReentrant ensure(deadline) ensurePath(path) returns (uint256[] memory amounts) {
         amounts = _swapTokensForExactETH(amountOut, amountInMax, path, to);
     }
 
     /**
-     * @notice A public function to sweep accidental ERC-20 transfers to this contract. Tokens are sent to admin (timelock)
+     * @notice A public function to sweep accidental ERC-20 transfers to this contract. Tokens are sent to address - to provided in input (timelock)
```
Comment (chechu) [inline contracts/Swap/SwapRouter.sol:744]:
```suggestion
     * @notice A public function to sweep accidental BEP-20 transfers to this contract. Tokens are sent to the address `to`, provided in input
```

File: contracts/Swap/SwapRouter.sol [hunk 2]
```diff
@@ -698,29 +789,54 @@ contract SwapRouter is Ownable2Step, RouterHelper, IPancakeSwapV2Router {
...
+        }
+    }
+
+    /**
+     * @notice Check is vToken listed in the pool.
+     * @param vToken Address of the vToken.
+     */
+    function _isVTokenListed(address vToken) internal view {
+        bool isListed = InterfaceComptroller(comptrollerAddress).markets(vToken);
+        if (isListed != true) {
```
Comment (chechu) [inline contracts/Swap/SwapRouter.sol:838]:
```suggestion
        if (!isListed) {
```

File: contracts/Swap/SwapRouter.sol [hunk 3]
```diff
@@ -81,19 +96,22 @@ contract SwapRouter is Ownable2Step, RouterHelper, IPancakeSwapV2Router {
...
         _swapExactTokensForTokens(amountIn, amountOutMin, path, address(this), TypesOfTokens.NON_SUPPORTING_FEE);
-        uint256 swapAmount = _getSwapAmount(path, balanceBefore);
-        _supply(path[path.length - 1], vTokenAddress, swapAmount);
+        uint256 swapAmount = _getSwapAmount(lastAsset, balanceBefore);
+        _supply(lastAsset, vTokenAddress, swapAmount);
     }
 
     /**
-     * @notice Swap token A for token B and supply to a Venus market
+     * @notice Swap Deflationary (a small amount of fee is deducted at the time of transfer of token) token A for token B and supply to a Venus market.
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/Swap/SwapRouter.sol]: Fixed, 346d32e59b64e7224302c6104f7c338fc7e38e60

File: contracts/Swap/SwapRouter.sol (hunk 2 above)
Comment (defcon022) [reply contracts/Swap/SwapRouter.sol]: Fixed, 346d32e59b64e7224302c6104f7c338fc7e38e60

**Review by chechu — APPROVED** [review]

**Review by GitGuru7 — APPROVED** [review]

**Review by chechu — APPROVED** [review]

**Pull request merged**